File: vripper-server/src/main/java/tn/mnlr/vripper/tasks/LinkScanRunnable.java
package tn.mnlr.vripper.tasks;

import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.NonNull;
import tn.mnlr.vripper.SpringContext;
import tn.mnlr.vripper.Utils;
import tn.mnlr.vripper.jpa.domain.LogEvent;
import tn.mnlr.vripper.jpa.domain.Queued;
import tn.mnlr.vripper.jpa.repositories.impl.LogEventRepository;
import tn.mnlr.vripper.services.PostService;
import tn.mnlr.vripper.services.SettingsService;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static tn.mnlr.vripper.jpa.domain.LogEvent.Status.ERROR;

@Slf4j
public class LinkScanRunnable implements Runnable {

  private static final Object LOCK = new Object();
  private final List<String> urlList;
  private final SettingsService settingsService;
  private final PostService postService;
  private final LogEventRepository eventRepository;
  private final LogEvent logEvent;

  public LinkScanRunnable(@NonNull List<String> urlList) {
    this.urlList = urlList;
    settingsService = SpringContext.getBean(SettingsService.class);
    postService = SpringContext.getBean(PostService.class);
    eventRepository = SpringContext.getBean(LogEventRepository.class);
    logEvent =
        new LogEvent(
            LogEvent.Type.SCAN,
            LogEvent.Status.PENDING,
            LocalDateTime.now(),
            "Links to scan:\n\t" + String.join("\n\t", urlList));
    eventRepository.save(logEvent);
  }

  @Override
  public void run() {
    synchronized (LOCK) {
      try {
        logEvent.setStatus(LogEvent.Status.PROCESSING);
        eventRepository.update(logEvent);
        ArrayList<Queued> queuedList = new ArrayList<>();
        List<String> unsupported = new ArrayList<>();
        List<String> unrecognized = new ArrayList<>();
        for (String url : urlList) {
          log.debug(String.format("Starting to process thread: %s", url));
          if (!url.startsWith(settingsService.getSettings().getVProxy())) {
            log.error(String.format("Unsupported link %s", url));
            unsupported.add(url);
            continue;
          }

          String threadId, postId;
          Matcher m =
              Pattern.compile(
                      Pattern.quote(settingsService.getSettings().getVProxy())
                          + "/threads/(\\d+)((.*p=)(\\d+))?")
                  .matcher(url);
          if (m.find()) {
            threadId = m.group(1);
            postId = m.group(4);
          } else {
            log.error(String.format("Cannot retrieve thread id from URL %s", url));
            unrecognized.add(url);
            continue;
          }
          queuedList.add(new Queued(url, threadId, postId));
        }
        StringBuilder errorMessage = new StringBuilder();
        if (!unsupported.isEmpty()) {
          errorMessage
              .append("Unsupported links:\n\t")
              .append(String.join("\n\t", unsupported))
              .append("\n\n");
        }
        if (!unrecognized.isEmpty()) {
          errorMessage
              .append("Unrecognized links:\n\t")
              .append(String.join("\n\t", unrecognized))
              .append("\n\n");
        }

        postService.processMultiPost(queuedList);
        if (!unsupported.isEmpty() || !unrecognized.isEmpty()) {
          logEvent.setStatus(LogEvent.Status.ERROR);
          logEvent.setMessage("Some links failed to be scanned: \n" + errorMessage);
        } else {
          logEvent.setStatus(LogEvent.Status.DONE);
        }
        eventRepository.update(logEvent);
      } catch (Exception e) {
        String error = "Error when scanning links";
        log.error(error, e);
        logEvent.setMessage(error + "\n" + Utils.throwableToString(e));
        logEvent.setStatus(ERROR);
        eventRepository.update(logEvent);
      }
    }
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/tasks/LinkScanRunnable.kt
package me.mnlr.vripper.tasks

import me.mnlr.vripper.SpringContext
import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.entities.LogEvent
import me.mnlr.vripper.entities.LogEvent.Status.*
import me.mnlr.vripper.formatToString
import me.mnlr.vripper.model.ThreadInput
import me.mnlr.vripper.repositories.impl.LogEventRepositoryImpl
import me.mnlr.vripper.services.SettingsService
import me.mnlr.vripper.services.ThreadPoolService
import java.time.LocalDateTime
import java.util.regex.Pattern

class LinkScanRunnable(private val urlList: List<String>) : Runnable {
    private val log by LoggerDelegate()
    private val settingsService: SettingsService =
        SpringContext.getBean(SettingsService::class.java)
    private val threadPoolService: ThreadPoolService =
        SpringContext.getBean(ThreadPoolService::class.java)
    private val eventRepository: LogEventRepositoryImpl =
        SpringContext.getBean(LogEventRepositoryImpl::class.java)
    private val logEvent: LogEvent

    init {
        logEvent = eventRepository.save(
            LogEvent(
                type = LogEvent.Type.SCAN,
                status = PENDING,
                time = LocalDateTime.now(),
                message = "Links to scan:\n\t${urlList.joinToString("\n\t")}"
            )
        )
    }

    override fun run() {
        try {
            eventRepository.update(logEvent.copy(status = PROCESSING))
            val threadInputList = mutableListOf<ThreadInput>()
            val unsupported = mutableListOf<String>()
            val unrecognized = mutableListOf<String>()
            for (link in urlList) {
                log.debug("Starting to process thread: $link")
                if (!link.startsWith(settingsService.settings.viperSettings.host)) {
                    log.error("Unsupported link $link")
                    unsupported.add(link)
                    continue
                }
                var threadId: String
                val m = Pattern.compile(
                    Pattern.quote(settingsService.settings.viperSettings.host) + "/threads/(\\d+).*"
                ).matcher(link)
                if (m.find()) {
                    threadId = m.group(1)
                } else {
                    log.error("Cannot retrieve thread id from URL $link")
                    unrecognized.add(link)
                    continue
                }
                threadInputList.add(ThreadInput(link = link, threadId = threadId))
            }
            val errorMessage: String = if (unsupported.isNotEmpty()) {
                """Unsupported links:
    ${unsupported.joinToString("\n\t")}
    """
            } else if (unrecognized.isNotEmpty()) {
                """Unrecognized links:
    ${unrecognized.joinToString("\n\t")}
    """
            } else {
                ""
            }
            if (errorMessage.isNotBlank()) {
                eventRepository.update(
                    logEvent.copy(
                        status = ERROR, message = "Some links failed to be scanned: \n$errorMessage"
                    )
                )
            }
            for (threadInput in threadInputList) {
//                if (thread.postId.isNotBlank()) {
//                    threadPoolService.generalExecutor.submit(
//                            SinglePostRunnable(
//                                thread.threadId, thread.postId
//                            )
//                        )
//                } else {
                threadPoolService.generalExecutor.submit(
                    ThreadLookupRunnable(
                        threadInput,
                        settingsService.settings
                    )
                )
//                }
            }
            eventRepository.update(logEvent.copy(status = DONE))
        } catch (e: Exception) {
            val error = "Error when scanning links"
            log.error(error, e)
            eventRepository.update(
                logEvent.copy(
                    status = ERROR, message = """
                $error
                ${e.formatToString()}
                """.trimIndent()
                )
            )
        }
    }
}